src/webgpu/shader/execution/expression/call/builtin/atomics/atomicCompareExchangeWeak.spec.ts

```diff
@@ -34,6 +37,264 @@ struct __atomic_compare_exchange_result<T> {
+    const outputExpected = new (typedArrayCtor(t.params.scalarKind))(bufferNumElements);
+    outputExpected.forEach((_, i) => (outputExpected[i] = i));
+    t.expectGPUBufferValuesEqual(outputBuffer, outputExpected);
+
+    // Exchanged buffer should be 'true' for every third value
+    const exchangedExpected = new (typedArrayCtor(t.params.scalarKind))(bufferNumElements);
+    exchangedExpected.forEach((_, i) => (exchangedExpected[i] = i % 3 === 0 ? 1 : 0));
+    t.expectGPUBufferValuesEqual(exchangedBuffer, exchangedExpected);
+
+    // And the input buffer should have been modified to a computed value
+    const inputExpected = new (typedArrayCtor(t.params.scalarKind))(bufferNumElements);
+    inputExpected.forEach(
+      (_, i) => (inputExpected[i] = i % 3 === 0 ? mapId.f(i * 2, numInvocations) : i)
+    );
+    t.expectGPUBufferValuesEqual(inputBuffer, inputExpected);
+  });
+
+g.test('compare_exchange_weak_workgroup')
+  .specURL('https://www.w3.org/TR/WGSL/#atomic-rmw')
+  .desc(
+    `
+AS is storage or workgroup
+T is i32 or u32
+
+fn atomicCompareExchangeWeak(atomic_ptr: ptr<AS, atomic<T>, read_write>, cmp: T, v: T) -> __atomic_compare_exchange_result<T>
+
+struct __atomic_compare_exchange_result<T> {
+  old_value : T,    // old value stored in the atomic
+  exchanged : bool, // true if the exchange was done
+}
+`
+  )
+  .params(u =>
+    u
+      .combine('workgroupSize', workgroupSizes)
+      .combine('dispatchSize', dispatchSizes)
+      .combine('mapId', keysOf(kMapId))
+      .combine('scalarKind', ['u32', 'i32'])
+  )
+  .fn(t => {
+    const numInvocations = t.params.workgroupSize;
+    const wgNumElements = numInvocations;
+    const scalarKind = t.params.scalarKind;
+    const dispatchSize = t.params.dispatchSize;
+    const mapId = kMapId[t.params.mapId];
+    const extra = mapId.wgsl(numInvocations, t.params.scalarKind); // Defines map_id()
+
+    const wgsl =
+      `
+      var<workgroup> wg: array<atomic<${scalarKind}>, ${wgNumElements}>;
+
+      @group(0) @binding(0)
+      var<storage, read_write> output: array<${scalarKind}, ${wgNumElements * dispatchSize}>;
+
+      @group(0) @binding(1)
+      var<storage, read_write> exchanged: array<${scalarKind}, ${wgNumElements * dispatchSize}>;
+
+      // Result of each workgroup is written to output[workgroup_id.x]
+      @group(0) @binding(2)
+      var<storage, read_write> wg_copy: array<${scalarKind}, ${wgNumElements * dispatchSize}>;
+
+      @compute @workgroup_size(${t.params.workgroupSize})
+      fn main(
+          @builtin(local_invocation_index) local_invocation_index: u32,
+          @builtin(workgroup_id) workgroup_id : vec3<u32>
+          ) {
+        let id = ${scalarKind}(local_invocation_index);
+        let global_id = ${scalarKind}(workgroup_id.x * ${wgNumElements} + local_invocation_index);
+
+        // Initialize wg[id] with this invocations global id
+        atomicStore(&wg[id], global_id);
+        workgroupBarrier();
```

Comment:
Strictly speaking, I think this barrier is unnecessary. Each invocation operates on a separate piece of data. Program ordering should ensure there is no reordering.

Reply:
Right, this is a vestige from other code I wrote where only invocation with `local_invocation_index == 0` would initialize all values. Will remove.

Reply:
Removed.